### PR TITLE
feat(manager/pipenv): Support custom environment variable usage in Pipfile source URLs 

### PR DESCRIPTION
**Pull request by fivetide**
## Changes

This change expands functionality of pipenv to allow for arbitrary credential placeholders as outlined in discussion #26447.
In addition, multiple source-urls can now be handled. 

## Context

- #26447 
- 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [X] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]
@fivetide thank you for this PR!

I found the logic flow a little hard to follow, and think this will be a little hard to maintain. I know you're mostly just extending the existing flow, but the additional complexity has now increased an existing problem of it being hard to understand.

Would you mind adding some comments to the flow so that it's easier to follow and for future maintainers? Sometimes improved variable names make commenting unnecessary too.

Once you're done with that, I'll see if there's anything I can add myself

File: lib/modules/manager/pipenv/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:34]:
```suggestion
```
mocked globally

File: lib/modules/manager/pipenv/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:21]: this reorder is probably the cause of failing ci.

wonder why it's imported twice 🤔

Comment (fivetide) [reply lib/modules/manager/pipenv/artifacts.spec.ts:20]: no clue, maybe intellij found it a great idea

File: lib/modules/manager/pipenv/artifacts.ts
Comment (rarkins) [inline lib/modules/manager/pipenv/artifacts.ts:192]: In theory this should never be reached, because the list of sourceUrls was already filtered using a parseUrl?

File: lib/modules/manager/pipenv/artifacts.ts
Comment (rarkins) [inline lib/modules/manager/pipenv/artifacts.ts:197]: Are these fallbacks to USERNAME and PASSWORD needed in any known scenarios or there "just in case"?

File: lib/modules/manager/pipenv/artifacts.ts
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:125]:
```suggestion
  }
  return undefined;
```
missing return?

File: lib/modules/manager/pipenv/artifacts.ts
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:135]: please return the parsed urls, so we don't need to parse multiple times (perf)

File: lib/modules/manager/pipenv/artifacts.ts
Comment (fivetide) [reply lib/modules/manager/pipenv/artifacts.ts:192]: correct. tsc just did not understand. that branch will be gone after next commit.

File: lib/modules/manager/pipenv/artifacts.ts
Comment (fivetide) [reply lib/modules/manager/pipenv/artifacts.ts:197]: initially as fallback, but now they are not needed anymore. will remove

File: lib/modules/manager/pipenv/artifacts.ts
Comment (fivetide) [reply lib/modules/manager/pipenv/artifacts.ts:125]: omitted intentionally. i didnt want to introduce a reason for the CI to demand coverage for an explicit statement. But lets see if it does...

File: lib/modules/manager/pipenv/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:1]:
```suggestion
```

File: lib/modules/manager/pipenv/artifacts.spec.ts
Comment (viceice) [reply lib/modules/manager/pipenv/artifacts.spec.ts:22]: - #28337

File: lib/modules/manager/pipenv/artifacts.ts
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:137]:
```suggestion
  return pipfile?.registryUrls
    .filter(is.urlInstance)
    .filter((url) => is.nonEmptyStringAndNotWhitespace(url.username)) ?? [];
```

File: lib/modules/manager/pipenv/artifacts.ts
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:147]:
```suggestion
  const match = regEx('([a-z0-9_]+)', 'i').exec(decodeURI(credential));
```
(perf)

File: lib/modules/manager/pipenv/artifacts.ts
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:156]:
```suggestion
  logger.trace(
```

you can remap via config if required

https://docs.renovatebot.com/configuration-options/#loglevelremap

File: lib/modules/manager/pipenv/artifacts.ts
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:186]:
```suggestion
    logger.trace(`Trying to add credentials for ${parsedSourceUrl}`);
```
Do we need the `tostring`?

File: lib/modules/manager/pipenv/artifacts.spec.ts
Comment (viceice) [reply lib/modules/manager/pipenv/artifacts.spec.ts:21]: please merge main to fix this

Comment (fivetide) [reply lib/modules/manager/pipenv/artifacts.spec.ts:21]: done :)

File: lib/modules/manager/pipenv/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:660]:
```suggestion
    expect(logger.warn).toHaveBeenCalledOnce();
```

File: lib/modules/manager/pipenv/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:641]: use jest it.each

File: lib/modules/manager/pipenv/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:640]: can you please use the tagged template syntax instead?

- https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#unit-testing

File: lib/modules/manager/pipenv/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:649]:
```suggestion
    'extracts correct environment variable from credential placeholder for `$credential`',
```

**Comment by rarkins** [comment]
@fivetide it's best that you don't merge from main unless it's to fix something. Otherwise it resets the tests and I need to manually approve it again. If/when @viceice it will start automerging, assuming tests have passed

**Comment by fivetide** [comment]
Hey @rarkins, @viceice  thanks for getting this PR ready. It looks to me that the merge is blocked by outdated changes. can you point me to how to resolve that?

**Comment by rarkins** [comment]
@fivetide it will now automerge once Michael approves

**Review by viceice — APPROVED** [review]

**Comment by fivetide** [comment]
Thanks a whole lot to you guys for helping me making this happen!

**Pull request merged**
